src/providers/store-provider/StoreProvider.tsx

```diff
@@ -0,0 +1,43 @@
+import { ReactElement, useCallback, useState } from 'react';
+
+import {
+  IStoreContext,
+  StoreContext,
+  StoreDispatchContext,
+} from './storeContext';
+
+interface IStoreProviderProps {
+  children: ReactElement;
+  initialValue: IStoreContext;
+  dispatchCallback?: (storeProperties: Partial<IStoreContext>) => void;
+}
+
+export const StoreProvider = ({
+  children,
+  initialValue,
+  dispatchCallback,
+}: IStoreProviderProps) => {
+  // Can be replaced with useReducer in future
+  const [store, setStore] = useState(initialValue);
+
+  const updateStoreProperties = useCallback(
+    (storeProperties: Partial<IStoreContext>) => {
+      setStore({
+        ...store,
+        ...storeProperties,
+      });
+      if (dispatchCallback) {
+        dispatchCallback(storeProperties);
+      }
+    },
+    [store, setStore, dispatchCallback]
+  );
```

Comment:
```suggestion
  const updateStoreProperties = useCallback(
    (storeProperties: Partial<IStoreContext>) => {
      setStore(prevStore => ({
        ...prevStore,
        ...storeProperties,
      }));
      if (dispatchCallback) {
        dispatchCallback(storeProperties);
      }
    },
    [setStore, dispatchCallback]
  );
```

Even better according to https://react.dev/reference/react/useCallback#updating-state-from-a-memoized-callback